File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
import sys
answer = 0
def dfs(word, cnt, visited, target, words):
    global answer
    
    if word == target:
        if answer==0:
            answer = cnt
        else:
            if cnt < answer:
                answer = cnt
        return
    
    for i in range(len(words)):
        if not visited[i]:
            cntt = 0
            for j in range(len(words[i])):
                if words[i][j] != word[j]:
                    cntt+=1
            if cntt == 1:
                visited[i] = True
                dfs(words[i], cnt+1, visited, target, words)
                visited[i] = False
                
    return  
def solution(begin, target, words):
    if len(words[0])!=len(begin) or len(begin)!=len(target):
        return 0
    
    visited = [False] * len(words)
    for i in range(len(words)):
        cnt = 0
        for j in range(len(words[i])):
            if words[i][j]!=begin[j]:
                cnt+=1
        if cnt == 1 and not visited[i]:
            visited[i] = True
            dfs(words[i], 1, visited, target, words)
            visited[i] = False
            
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    static int result;
    public int solution(String begin, String target, String[] words) {
        result = Integer.MAX_VALUE;
        
        // 만약 타겟 단어가 words 배열에 없다면
        if(!check(words, target)) return 0;
        
        dfs(0, words, begin, target, new boolean[words.length]);
        
        if (result==Integer.MAX_VALUE) result=0;
        return result;
    }
    
    static boolean check(String[] words, String target) {
        for (int i=0;i<words.length;i++) {
            if (words[i].equals(target)) return true;
        }
        return false;
    }
    
    static boolean checkOne(String begin, String target) {
        int cnt = 0;
        for (int i=0;i<begin.length();i++) {
            if (begin.charAt(i)!=target.charAt(i)) cnt++;
        }
        if(cnt==1) return true;
        else return false;
    }
    
    static void dfs(int cnt, String[] words, String now, String target, boolean[] visited) {
        if (now.equals(target)) {
            result = Math.min(result, cnt);
            return;
        }
        
        if (cnt==words.length) return;
        for (int i=0;i<words.length;i++) {
            if (!visited[i]) {
                if (checkOne(now, words[i])) {
                    visited[i] = true;
                    dfs(cnt+1, words, words[i], target, visited);
                    visited[i] = false;
                }
            }
        }
    }
}